plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java

```diff
@@ -0,0 +1,178 @@
+    {
+        RowBlock rowIdBlock = (RowBlock) page.getBlock(columnValueAndRowIdChannels.get(columnValueAndRowIdChannels.size() - 1));
+        deleteRows(rowIdBlock);
+        Map<Integer, Block> updatedColumnBlocks = new HashMap<>();
+        for (int i = 0; i < columnValueAndRowIdChannels.size() - 1; i++) {
+            IcebergColumnHandle updatedColumn = updateColumns.get(i);
+            Block updatedValues = page.getBlock(columnValueAndRowIdChannels.get(i));
+            updatedColumnBlocks.put(updatedColumn.getId(), updatedValues);
+        }
+
+        Block[] updatedRows = new Block[allTableColumns.size()];
+        Block[] oldRows = ((RowBlock) rowIdBlock.getRawFieldBlocks()[2]).getRawFieldBlocks();
+        for (int i = 0; i < allTableColumns.size(); i++) {
+            int columnId = allTableColumns.get(i).getId();
+            if (updatedColumnBlocks.containsKey(columnId)) {
+                updatedRows[i] = updatedColumnBlocks.get(columnId);
+            }
+            else {
+                updatedRows[i] = oldRows[i];
+            }
+        }
+
+        updateRowSink.appendPage(new Page(page.getPositionCount(), updatedRows));
+    }
+
+    @Override
+    public CompletableFuture<Collection<Slice>> finish()
+    {
+        try {
+            Collection<Slice> slices = posDeleteSink.finish().get();
+            slices.addAll(updateRowSink.finish().get());
+            return CompletableFuture.completedFuture(slices);
+        }
+        catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        Page sourcePage = source.getNextPage();
+        if (sourcePage == null) {
+            return null;
+        }
+        Block[] rowIdComponentBlocks = new Block[3];
+        int channelCount = sourcePage.getChannelCount();
+        int pageSize = sourcePage.getPositionCount();
+        // file_path
+        rowIdComponentBlocks[0] = RunLengthEncodedBlock.create(VARCHAR, filePathSlice, pageSize);
+        rowIdComponentBlocks[1] = sourcePage.getBlock(channelCount - 2);
+        Block[] fieldBlocks = new Block[allTableColumns.size()];
+        for (int i = 0; i < allTableColumns.size(); i++) {
+            fieldBlocks[i] = sourcePage.getBlock(i);
+        }
+        rowIdComponentBlocks[2] = RowBlock.fromFieldBlocks(pageSize, Optional.empty(), fieldBlocks);
+
+        Block[] resultBlocks = new Block[queriedColumns.size()];
+        for (int i = 0; i < queriedColumns.size(); i++) {
+            IcebergColumnHandle columnHandle = queriedColumns.get(i);
+            if (ROW_ID_COLUMN_NAME.equals(columnHandle.getName())) {
+                resultBlocks[i] = RowBlock.fromFieldBlocks(pageSize, Optional.empty(), rowIdComponentBlocks);
+            }
+            else {
+                resultBlocks[i] = sourcePage.getBlock(allTableColumns.indexOf(columnHandle));
```

Comment:
`indexOf` use here looks quadratic, and we seem to be doing this for every page.